### PR TITLE
GEODE-6702: optimize equals by not calling compareTo

### PR DESCRIPTION
**Pull request by dschneider-pivotal**
The compareTo implementation needs to get the InetAddress as a byte array
which causes to byte array allocations.
By changing InternalDistributedMember.equals to not call compareTo InetAddress.equals
can be used instead which prevents the byte array allocations.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/distributed/internal/membership/InternalDistributedMember.java [hunk 1]
```diff
@@ -637,7 +638,68 @@ public boolean equals(Object obj) {
...
-    return compareTo((InternalDistributedMember) obj) == 0;
+    InternalDistributedMember other = (InternalDistributedMember) obj;
+
+    int myPort = getPort();
+    int otherPort = other.getPort();
+    if (myPort != otherPort) {
+      return false;
+    }
+
+    InetAddress myAddr = getInetAddress();
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/membership/InternalDistributedMember.java]: Could we replace may of these blocks of code with.
```
if (!Object.equals(getInetAddress(), other.getInetAddress())) {
  return false;
}

Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/membership/InternalDistributedMember.java]: That is a good idea. We would not want to use it on primitive fields (like the ports) and the inetAddress comparison has a special case that if they are both null then we return true immediately. I will make the change to use Objects.equals where possible and update this PR.

**Comment by dschneider-pivotal** [comment]
> Won't the runtime optimize the compareTo call?

I don't know. I know that compareTo has to ask the InetAddress for a byte array representation of the address so that it can figure out if one address is greater or lessor than the other. It was these byte array allocations that showed up in the profiler. I think this would be garbage that the collector has to deal with. The new equals logic also has fewer conditions that it needs to test for but that was not the reason for this pull request and I doubt it would be much of a difference. If I could have found a way to change compareTo to not do the byte array allocations (by calling InetAddress.getAddress) I would have made that change add kept equals implemented by calling compareTo.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/membership/InternalDistributedMember.java [hunk 2]
```diff
@@ -637,7 +639,51 @@ public boolean equals(Object obj) {
...
+    InetAddress otherAddr = other.getInetAddress();
+    if (myAddr == null && otherAddr == null) {
+      return true;
+    } else if (!Objects.equals(myAddr, otherAddr)) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/membership/InternalDistributedMember.java]: Object.equals handles the `null` case as well.

Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/membership/InternalDistributedMember.java]: I don't think so. Objects.equals(null, null) returns true. But so does other non-null InetAddresses. But we are only willing to stop here if either they are not equal OR they are both null. If the addresses are non-null and equal then we continue on and compare other things. This logic was derived from the compareTo logic that I did not want to change. I'm not sure why it did not treat "null, null" as equal and just continue to compare other things.

Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/membership/InternalDistributedMember.java]: Oh I see... yeah that is odd.

**Pull request merged**